hazelcast/src/main/java/com/hazelcast/sql/impl/SqlInternalService.java

```diff
@@ -0,0 +1,194 @@
+import com.hazelcast.sql.impl.state.QueryStateRegistryUpdater;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+
+/**
+ * Proxy for SQL service. Backed by either Calcite-based or no-op implementation.
+ */
+public class SqlInternalService {
+    /** Memory assigned to a single edge mailbox. Will be reworked to dynamic mode when memory manager is implemented. */
+    private static final long MEMORY_PER_EDGE_MAILBOX = 512 * 1024;
+
+    /** Default flow control factory. */
+    private static final FlowControlFactory FLOW_CONTROL_FACTORY = SimpleFlowControlFactory.INSTANCE;
+
+    /** Node service provider. */
+    private final NodeServiceProvider nodeServiceProvider;
+
+    /** Registry for running queries. */
+    private final QueryStateRegistry stateRegistry;
+
+    /** Operation manager. */
+    private final QueryOperationHandlerImpl operationHandler;
+
+    /** State registry updater. */
+    private final QueryStateRegistryUpdater stateRegistryUpdater;
+
+    public SqlInternalService(
+        String instanceName,
+        NodeServiceProvider nodeServiceProvider,
+        InternalSerializationService serializationService,
+        int operationThreadCount,
+        int fragmentThreadCount,
+        int outboxBatchSize,
+        long stateCheckFrequency
+    ) {
+        this.nodeServiceProvider = nodeServiceProvider;
+
+        // Create state registries since they do not depend on anything.
+        stateRegistry = new QueryStateRegistry(nodeServiceProvider);
+
+        // Operation handler depends on state registry.
+        operationHandler = new QueryOperationHandlerImpl(
+            instanceName,
+            nodeServiceProvider,
+            serializationService,
+            stateRegistry,
+            outboxBatchSize,
+            FLOW_CONTROL_FACTORY,
+            fragmentThreadCount,
+            operationThreadCount
+        );
+
+        // State checker depends on state registries and operation handler.
+        stateRegistryUpdater = new QueryStateRegistryUpdater(
+            instanceName,
+            nodeServiceProvider,
+            stateRegistry,
+            operationHandler,
+            stateCheckFrequency
+        );
+    }
+
+    public void start() {
+        stateRegistryUpdater.start();
+    }
+
+    public void reset() {
+        stateRegistry.reset();
+    }
+
+    public void shutdown() {
+        stateRegistryUpdater.stop();
+        operationHandler.stop();
+
+        reset();
+    }
+
+    /**
+     * Internal query execution routine.
+     *
+     * @return Query state.
+     */
+    public QueryState execute(Plan plan, List<Object> params, long timeout, int pageSize) {
+        // Get local member ID and check if it is still part of the plan.
+        UUID localMemberId = nodeServiceProvider.getLocalMemberId();
+
+        if (!plan.getPartitionMap().containsKey(localMemberId)) {
+            throw QueryException.memberLeave(localMemberId);
```

Comment:
Do we plan to handle such sort of exceptions somewhere on the higher level to hide it from the user?

Reply:
Yes, all member exceptions are propagated upwards and will be handled similarly in two places - API entry point and cursor returned from API. You may see it in action in the prototype branch. This is not part of the current PR, because we do not have API yet.